pkg/api/defaults.go

```diff
@@ -508,6 +512,101 @@ func (p *Properties) setHostedMasterProfileDefaults() {
 	p.HostedMasterProfile.Subnet = DefaultKubernetesMasterSubnet
 }
 
+func (p *Properties) setCustomCloudProfileDefaults() {
+
+	if p.IsAzureStackCloud() {
+
+		//azureStackCloudSpec is the default configurations for azure stack with public Azure.
+		azureStackCloudSpec := AzureEnvironmentSpecConfig{
+			CloudName: AzureStackCloud,
+			//DockerSpecConfig specify the docker engine download repo
+			DockerSpecConfig: DefaultDockerSpecConfig,
+			//KubernetesSpecConfig is the default kubernetes container image url.
+			KubernetesSpecConfig: DefaultKubernetesSpecConfig,
+			DCOSSpecConfig:       DefaultDCOSSpecConfig,
+			EndpointConfig: AzureEndpointConfig{
+				ResourceManagerVMDNSSuffix: "",
+			},
+			OSImageConfig: map[Distro]AzureOSImageConfig{
+				Ubuntu:          DefaultUbuntuImageConfig,
+				RHEL:            DefaultRHELOSImageConfig,
+				CoreOS:          DefaultCoreOSImageConfig,
+				AKS:             DefaultAKSOSImageConfig,
+				AKSDockerEngine: DefaultAKSDockerEngineOSImageConfig,
+			},
+		}
+
+		//Set default value for ResourceManagerVMDNSSuffix
+		azureStackCloudSpec.EndpointConfig.ResourceManagerVMDNSSuffix = p.CustomCloudProfile.Environment.ResourceManagerVMDNSSuffix
+
+		// Use the custom input to overwrite the default values in AzureStackCloudSpec
+		if p.CustomCloudProfile.AzureEnvironmentSpecConfig != nil {
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.CloudName != "" {
+				azureStackCloudSpec.CloudName = p.CustomCloudProfile.AzureEnvironmentSpecConfig.CloudName
+			}
+
+			// DockerSpecConfig
+			if p.CustomCloudProfile.AzureEnvironmentSpecConfig.DockerSpecConfig.DockerComposeDownloadURL != "" {
```

Comment:
Can we have a helper method which checks if a string is nil and assigns a default value then? The default value will also be passed as a parameter to the helper method. @honcao

Reply:
@tariq1890 , sure, PR is updated